docs/source/design/python_imports.rst

```diff
@@ -0,0 +1,65 @@
+.. -*- mode: rst -*-
+.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:
+
+.. _chap_design_python_imports:
+
+************************
+Python import statements
+************************
+
+.. topic:: Specification scope and status
+
+   This specification describes the current (albeit incomplete) implementation.
+
+The following rules apply to any ``import`` statement in the code base:
```

Comment:
brief overview of various "mainstream" libraries on use of relative imports

<details>
<summary>cpython: good amount of local, or one level up, but (thank god) not that many for higher up imports,majority of imports absolute (but didn't check if used where local could be used); tests seems to use abs imports</summary> 

```shell
$> git grep 'from \..* import' | nl | tail       
   756	Tools/c-analyzer/c_parser/preprocessor/pure.py:from ..source import (
   757	Tools/c-analyzer/c_parser/preprocessor/pure.py:from . import common as _common
   758	Tools/c-analyzer/cpython/__main__.py:from . import _analyzer, _capi, _files, _parser, REPO_ROOT
   759	Tools/c-analyzer/cpython/_analyzer.py:from . import _parser, REPO_ROOT
   760	Tools/c-analyzer/cpython/_capi.py:from ._files import iter_header_files, resolve_filename
   761	Tools/c-analyzer/cpython/_capi.py:from . import REPO_ROOT
   762	Tools/c-analyzer/cpython/_files.py:from . import REPO_ROOT, INCLUDE_DIRS, SOURCE_DIRS
   763	Tools/c-analyzer/cpython/_parser.py:from . import REPO_ROOT
   764	Tools/peg_generator/pegen/keywordgen.py:from .build import build_parser, generate_token_definitions
   765	Tools/peg_generator/pegen/keywordgen.py:from .c_generator import CParserGenerator

$> git grep 'from \.\..* import' | nl | tail
   204	Misc/HISTORY:- The compiler will report a SyntaxError if "from ... import *" occurs
   205	Misc/HISTORY:There's also a variant for 'from ... import':
   206	Misc/HISTORY:emulating from ... import *.
   207	Misc/NEWS.d/3.7.0a1.rst:Set the 'path' and 'name' attribute on ImportError for ``from ... import
   208	Misc/NEWS.d/3.8.0a1.rst:Importing names from already imported module with "from ... import ..." is
   209	Tools/c-analyzer/c_parser/parser/__init__.py:from ..info import ParsedItem
   210	Tools/c-analyzer/c_parser/parser/_info.py:from ..info import KIND, ParsedItem, FileInfo
   211	Tools/c-analyzer/c_parser/preprocessor/__init__.py:from ..source import (
   212	Tools/c-analyzer/c_parser/preprocessor/common.py:from ..info import FileInfo, SourceLine
   213	Tools/c-analyzer/c_parser/preprocessor/pure.py:from ..source import (

$> git grep 'from [^.].* import' | nl | tail 
  5705	Tools/unittestgui/unittestgui.py:from tkinter import simpledialog
  5706	setup.py:from glob import glob, escape
  5707	setup.py:    from distutils.command.build_ext import build_ext
  5708	setup.py:    from distutils.command.build_scripts import build_scripts
  5709	setup.py:    from distutils.command.install import install
  5710	setup.py:    from distutils.command.install_lib import install_lib
  5711	setup.py:    from distutils.core import Extension, setup
  5712	setup.py:    from distutils.errors import CCompilerError, DistutilsError
  5713	setup.py:    from distutils.spawn import find_executable
  5714	setup.py:        from os.path import join, exists

```
</details>

<details>
<summary>numpy: a good amount of "semantically bound/local" relative imports, tests do not use relative imports despite seems to be consistently located through the hierarchy</summary> 

```shell
(git)lena:~/proj/numpy[master]git
$> git grep 'from \.' | nl | tail
   359	numpy/typing/_ufunc.pyi:from ._scalars import _ScalarLike_co
   360	numpy/typing/_ufunc.pyi:from ._array_like import ArrayLike, _ArrayLikeBool_co, _ArrayLikeInt_co
   361	numpy/typing/_ufunc.pyi:from ._dtype_like import DTypeLike
   362	numpy/version.py:from ._version import get_versions
   363	tools/npy_tempita/__init__.py:from ._looper import looper
   364	tools/npy_tempita/__init__.py:from .compat3 import (
   365	tools/npy_tempita/_looper.py:from .compat3 import basestring_
   366	tools/openblas_support.py:                        warnings.warn("loaded more than 1 DLL from .libs:"
   367	versioneer.py:    from ._version import get_versions
   368	versioneer.py:from ._version import get_versions
(dev3) 1 15280 [2].....................................:Wed 18 Aug 2021 12:24:37 PM EDT:.
(git)lena:~/proj/numpy[master]git
$> git grep 'from \.\.'          
numpy/testing/_private/nosetester.py:                from ...distutils import cpuinfo
(dev3) 1 15281 [2].....................................:Wed 18 Aug 2021 12:25:00 PM EDT:.
(git)lena:~/proj/numpy[master]git
$> find -iname tests
./numpy/core/tests
./numpy/distutils/tests
./numpy/f2py/tests
./numpy/lib/tests
./numpy/linalg/tests
./numpy/random/tests
./numpy/testing/tests
./numpy/fft/tests
./numpy/tests
./numpy/ma/tests
./numpy/compat/tests
./numpy/matrixlib/tests
./numpy/polynomial/tests
./numpy/typing/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/lib/python3.7/site-packages/Cython/Compiler/Tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/lib/python3.7/site-packages/Cython/Build/Tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/lib/python3.7/site-packages/Cython/Tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/lib/python3.7/site-packages/Cython/Debugger/Tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/doc/sphinxext/numpydoc/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/compat/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/core/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/distutils/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/f2py/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/fft/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/lib/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/linalg/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/ma/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/matrixlib/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/polynomial/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/random/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/testing/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/tests
(dev3) 1 15282 [2].....................................:Wed 18 Aug 2021 12:25:21 PM EDT:.
(git)lena:~/proj/numpy[master]git
$> git grep 'from \.' | grep tests/ | nl | tail
    14	numpy/f2py/tests/test_regression.py:from . import util
    15	numpy/f2py/tests/test_return_character.py:from . import util
    16	numpy/f2py/tests/test_return_complex.py:from . import util
    17	numpy/f2py/tests/test_return_integer.py:from . import util
    18	numpy/f2py/tests/test_return_logical.py:from . import util
    19	numpy/f2py/tests/test_return_real.py:from . import util
    20	numpy/f2py/tests/test_semicolon_split.py:from . import util
    21	numpy/f2py/tests/test_size.py:from . import util
    22	numpy/f2py/tests/test_string.py:from . import util
    23	numpy/lib/tests/test_histograms.py:            rec = sup.record(RuntimeWarning, 'Converting input from .*')
```
</details>

<details>
<summary>pandas: some amount of local imports, tests included, but majority are absolute</summary> 

```shell
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from \.\.\..* import' | nl | tail
(dev3) 1 15311 [2].....................................:Wed 18 Aug 2021 12:34:12 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from \..* import' | nl | tail 
   132	pandas/tests/indexes/ranges/test_range.py:from ..test_numeric import Numeric
   133	pandas/tests/indexes/timedeltas/test_timedelta.py:from ..datetimelike import DatetimeLike
   134	pandas/tests/tseries/offsets/test_fiscal.py:from .common import assert_is_on_offset, assert_offset_equal
   135	pandas/tests/tseries/offsets/test_fiscal.py:from .test_offsets import Base, WeekDay
   136	pandas/tests/tseries/offsets/test_offsets.py:from .common import assert_is_on_offset, assert_offset_equal
   137	pandas/tests/tseries/offsets/test_ticks.py:from .common import assert_offset_equal
   138	pandas/tests/tseries/offsets/test_yqm_offsets.py:from .common import assert_is_on_offset, assert_offset_equal
   139	pandas/tests/tseries/offsets/test_yqm_offsets.py:from .test_offsets import Base
   140	versioneer.py:    from ._version import get_versions
   141	versioneer.py:from ._version import get_versions
(dev3) 1 15312 [2].....................................:Wed 18 Aug 2021 12:34:17 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from \..* import' | nl | grep -v tests/ | tail
    79	pandas/_libs/tslibs/offsets.pyx:from .timedeltas import Timedelta
    80	pandas/_libs/tslibs/offsets.pyx:from .timestamps import Timestamp
    81	pandas/_libs/tslibs/offsets.pyx:        from .vectorized import normalize_i8_timestamps
    82	pandas/_libs/tslibs/tzconversion.pyx:        from .timedeltas import delta_to_nanoseconds
    83	pandas/_libs/tslibs/vectorized.pyx:from .dtypes import Resolution
    84	pandas/core/arrays/boolean.py:from .masked import BaseMaskedArray, BaseMaskedDtype
    85	pandas/core/arrays/integer.py:from .masked import BaseMaskedArray, BaseMaskedDtype
    86	pandas/core/arrays/timedeltas.py:        from .period import PeriodArray
   140	versioneer.py:    from ._version import get_versions
   141	versioneer.py:from ._version import get_versions
(dev3) 1 15313 [2].....................................:Wed 18 Aug 2021 12:34:30 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from [^.].* import' | nl | tail             
  4547	setup.py:    from distutils.command.build_ext import build_ext as _build_ext
  4548	versioneer.py:    # keywords. When used from setup.py, we don't want to import _version.py,
  4549	versioneer.py:    # keywords. When used from setup.py, we don't want to import _version.py,
  4550	versioneer.py:from warnings import catch_warnings
  4551	versioneer.py:    from distutils.core import Command
  4552	versioneer.py:    from distutils.command.build_py import build_py as _build_py
  4553	versioneer.py:        from cx_Freeze.dist import build_exe as _build_exe
  4554	versioneer.py:        from setuptools.command.sdist import sdist as _sdist
  4555	versioneer.py:        from distutils.command.sdist import sdist as _sdist
  4556	web/pandas/community/ecosystem.md:saved/loaded to/from a file. Spyder can also import data from a variety
(dev3) 1 15314 [2].....................................:Wed 18 Aug 2021 12:34:48 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from [^.].* import' | nl | grep tail
'   389	pandas/_libs/testing.pyx:                from pandas._testing import raise_assert_detail
   391	pandas/_libs/testing.pyx:            from pandas._testing import raise_assert_detail
   392	pandas/_libs/testing.pyx:            from pandas._testing import raise_assert_detail
  2973	pandas/tests/frame/methods/test_head_tail.py:from pandas import DataFrame
(dev3) 1 15315 [2].....................................:Wed 18 Aug 2021 12:34:54 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from [^.].* import' | nl | grep 'tests/' | tail
  4473	pandas/tests/window/test_pairwise.py:from pandas import DataFrame, MultiIndex, Series, date_range
  4474	pandas/tests/window/test_pairwise.py:from pandas.core.algorithms import safe_sort
  4475	pandas/tests/window/test_rolling.py:from datetime import datetime, timedelta
  4476	pandas/tests/window/test_rolling.py:from pandas.errors import UnsupportedFunctionCall
  4477	pandas/tests/window/test_rolling.py:from pandas import DataFrame, Series, date_range
  4478	pandas/tests/window/test_rolling.py:from pandas.core.window import Rolling
  4479	pandas/tests/window/test_timeseries_window.py:from pandas import (
  4480	pandas/tests/window/test_window.py:from pandas.errors import UnsupportedFunctionCall
  4481	pandas/tests/window/test_window.py:from pandas import Series
  4482	pandas/tests/window/test_window.py:from pandas.core.window import Window

```
</details>

Reply:
Thanks for the empirical data! This seems to be fairly in line with the description we have in the document.